assets/js/util/tracking/index.js

```diff
@@ -0,0 +1,194 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const DEFAULT_CONFIG = {
+	isFirstAdmin: false,
+	trackingEnabled: false,
+	trackingID: '',
+	referenceSiteURL: '',
+	userIDHash: '',
+};
+
+/**
+ * Tracking configuration.
+ */
+const config = { ...DEFAULT_CONFIG };
+
+/**
+ * The promise created when tracking is enabled.
+ * This reference is kept and returned for subsequent calls.
+ */
+let enableTrackingPromise;
+
+/**
+ * gtag script identifier.
+ */
+const SCRIPT_IDENTIFIER = 'data-googlesitekit-gtag';
+
+/**
+ * Data layer global used for internal/private Site Kit data.
+ */
+export const DATA_LAYER = '_googlesitekitDataLayer';
+
+/**
+ * Initializes tracking.
+ *
+ * @param {Object} _global The global object. Optional. (Used for testing only)
+ * @param {boolean} _resetPromise Clears the internal reference to the enableTrackingPromise (Optional - for testing only)
+ */
+export async function bootstrapTracking( _global = global, _resetPromise = false ) {
+	const {
+		isFirstAdmin,
+		trackingID,
+		trackingEnabled,
+		referenceSiteURL,
+		userIDHash,
+	} = _global._googlesitekitBase || DEFAULT_CONFIG;
+
+	Object.assign( config, DEFAULT_CONFIG, {
+		isFirstAdmin,
+		trackingID,
+		trackingEnabled,
+		userIDHash,
+		// Remove any trailing slash from the reference URL.
+		referenceSiteURL: referenceSiteURL.toString().replace( /\/+$/, '' ),
+	} );
+
+	if ( _resetPromise ) {
+		enableTrackingPromise = null;
+	}
+
+	return toggleTracking( trackingEnabled, _global );
+}
+
+/**
+ * Change the active state of tracking.
+ *
+ * @param {boolean} activeStatus The new state to set.
+ * @param {Object} _global The global object. Optional. (Used for testing only)
+ * @return {Promise} A promise that resolves with an object { trackingEnabled: (bool) }.
+ */
+export async function toggleTracking( activeStatus, _global = global ) {
+	if ( !! activeStatus ) {
+		return enableTracking( _global );
+	}
+
+	return disableTracking();
+}
+
+/**
+ * Enables tracking by injecting the necessary script tag if not present.
+ *
+ * @return {Promise} A promise that resolves when the script is loaded and ready.
+ * @param {Object} _global The global object. Optional. (Used for testing only)
+ */
+export async function enableTracking( _global = global ) {
+	// Return the existing promise if already called.
+	if ( enableTrackingPromise ) {
+		return enableTrackingPromise;
+	}
+
+	config.trackingEnabled = true;
+
+	dataLayerPush( _global, 'js', new Date() );
+	dataLayerPush( _global, 'config', config.trackingID );
+
+	const { document } = _global;
```

Comment:
This should really come from `global`. We don't need to go as far as support custom mocks for `document`.

Reply:
See @tofumatt 's comment here https://github.com/google/site-kit-wp/pull/1079#discussion_r371776734

Reply:
Okay, that's fair. One thing we shouldn't do though (as mentioned in https://github.com/google/site-kit-wp/pull/1079#discussion_r373583361) is not add this kind of arguments in public or high-level API functions. This is very low-level functionality and should be handled in a low-level API.

Although we don't currently plan exposing these functions, they are still the "public" functions of this module and only should include minimum parameters that are needed for the primary purpose.